engine/render/src/render/buffer.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright 2009-2014 Ragnar Svensson, Christian Murray
+// Licensed under the Defold License version 1.0 (the "License"); you may not use
+// this file except in compliance with the License.
+//
+// You may obtain a copy of the License, together with FAQs at
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "render_private.h"
+
+namespace dmRender
+{
+    static HRenderBuffer NewRenderBuffer(dmGraphics::HContext graphics_context, RenderBufferType type)
+    {
+        switch(type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER: return (HRenderBuffer) dmGraphics::NewVertexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:  return (HRenderBuffer) dmGraphics::NewIndexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+            default:assert(0);
+        }
+        return (HRenderBuffer) 0;
+    }
+
+    static void DeleteRenderBuffer(HRenderBuffer buffer_to_remove, RenderBufferType type)
+    {
+        switch(type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
+                dmGraphics::DeleteVertexBuffer((dmGraphics::HVertexBuffer) buffer_to_remove);
+                break;
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
+                dmGraphics::DeleteIndexBuffer((dmGraphics::HIndexBuffer) buffer_to_remove);
+                break;
+            default:assert(0);
+        }
+    }
+
+    static inline void CreateAndPush(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        buffer->m_Buffers.Push(NewRenderBuffer(render_context->m_GraphicsContext, buffer->m_Type));
+    }
+
+    static inline void PopAndDelete(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        DeleteRenderBuffer(buffer->m_Buffers.Back(), buffer->m_Type);
+        buffer->m_Buffers.Pop();
+    }
+
+    HBufferedRenderBuffer NewBufferedRenderBuffer(HRenderContext render_context, RenderBufferType type)
+    {
+        BufferedRenderBuffer* buffer = new BufferedRenderBuffer();
+        memset(buffer, 0, sizeof(BufferedRenderBuffer));
+        RewindBuffer(render_context, buffer);
+        buffer->m_Type = type;
+        return buffer;
+    }
+
+    void DeleteBufferedRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        for (int i = 0; i < buffer->m_Buffers.Size(); ++i)
+        {
+            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
+        }
+        delete buffer;
+    }
+
+    HRenderBuffer AdvanceRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
```

Comment:
I think the pattern of NewRenderBuffer() + AdvanceRenderBuffer() is a bit unusua imho.
I'd expect the buffer to either have one item already after New(), or that Advance just skip to the next index.

Perhaps something name it like AddRenderBuffer() ?

Reply:
Yes, I was running into this in my head as well when developing, I couldn't find a good name for what the function actually did, since I want a separation of which of the internal buffers are "current" and a way of adding new buffers. But AddRenderBuffer is a good choice I think :)